docs/html/user_guide.rst

```diff
@@ -1306,6 +1306,238 @@ issue tracker`_ if you believe that your problem has exposed a bug in pip.
+``pip install`` process. What pip is trying to do is complicated - it is
+working through potentially millions of package versions to identify the
+comptible versions.
+
+There is no guaranteed solution to backtracking but you can reduce it -
+here are a number of ways.
+
+.. _1-allow-pip-to-complete-its-backtracking:
+
+1. Allow pip to complete its backtracking
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In most cases, pip will complete the backtracking process successfully.
+It is possible this could take a very long time to complete - this may
+not be the preferred option.
+
+However there is a possibility pip will not be able to find a set of
+compatible versions.
+
+If you'd prefer not to wait, you can interrupt pip (ctrl and c) and use
+constraints to reduce the number of package versions it tries.
+
+.. _2-reduce-the-versions-of-the-backtracking-package:
+
+2. Reduce the versions of the backtracking package
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If pip is backtracking more than you'd like, the next option is to
+constrain the number of package versions it tries.
+
+A first good candidate for this constraining is the package(s) it is
+backtracking on (e.g. in the above example - ``cup``).
+
+You could try:
+
+``pip install tea cup > 3.13``
+
+This will reduce the number of versions of ``cup`` it tries, and
+possibly reduce the time pip takes to install.
+
+There is a possibility that if you're wrong (in this case a newer
+version would have worked) then you missed the chance to use it. This
+can be trial and error.
+
+.. _3-use-constraint-files-or-lockfiles:
+
+3. Use constraint files or lockfiles
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This option is a progression of 2 above. It requires users to know how
+to inspect:
+
+-  the packages they're are trying to install
+-  the package release frequency and compatibility policies
+-  their release notes and changelogs from past versions
+
+During deployment creating a lockfile stating the exact package and
+version number for for each dependency of that package. You can do this
+with `pip-tools <https://github.com/jazzband/pip-tools/>`__.
+
+This means the "work" is done once during development process, and so
+will save users this work during deployment.
+
+The pip team is not available to provide support in helping you create a
+suitable constraints file.
+
+.. _4-be-more-strict-on-package-dependencies-during-development:
+
+4. Be more strict on package dependencies during development
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+For package maintainers during the development, give pip some help by
+creating constraint files for the dependency tree. This will reduce the
+number of versions it will try.
+
+If you (as the user installing a package) see the package has specific
+version dependencies, you could contact the maintainer and ask them to
+be more strict with the packages requirements.
```

Comment:
In general, stricter dependencies on general-purpose libraries an anti-pattern, and is *not* advised. It makes it much harder to use libraries, and increases the possibility that no resolution is possible.

I'm pretty strongly against this suggestion, as it could create a culture of over-specification.

Reply:
OK thanks. I'll defer to your experience on this part.